File: hot100/53. 最大子数组和.py
# https://leetcode-cn.com/problems/maximum-subarray/
import sys
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        pre = - sys.maxsize
        ret = pre
        for n in nums:
            pre = max(pre+n, n)
            if pre > ret:
                ret = pre
        return ret


x = Solution()
print(x.maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))

File: ts-version/53. 最大子数组和.ts
function maxSubArray(nums: number[]): number {
    let sum = 0;
    let max = -Infinity;

    for (const num of nums) {
        if (num > sum + num) {
            // 当前数字+前序 < 当前数字
            sum = 0;
        }
        sum += num;
        max = max < sum ? sum : max;
    }

    return max;
}
